crates/ra_hir/src/ty.rs

```diff
@@ -1042,12 +1046,24 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
             },
             Expr::Tuple { exprs } => {
                 let mut ty_vec = Vec::with_capacity(exprs.len());
-                for arg in exprs.iter() {
-                    ty_vec.push(self.infer_expr(*arg, &Expectation::none())?);
+                for expr in exprs.iter() {
+                    ty_vec.push(self.infer_expr(*expr, &Expectation::none())?);
                 }
 
                 Ty::Tuple(Arc::from(ty_vec))
             }
+            Expr::Array { exprs } => {
+                let mut ty = Ty::Unknown;
+                for expr in exprs.iter() {
+                    let infered = self.infer_expr(*expr, &Expectation::none())?;
+                    match infered {
+                        Ty::Unknown => (),
+                        _ => ty = infered,
+                    }
```

Comment:
you're not actually making sure all the elements have the same type anywhere here. The best way to do that would be to pass the element type as expected type (so `Expectation::has_type(ty)`).